src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs

```diff
@@ -736,6 +738,110 @@ private static bool TryCompareTimeZoneInformationToRegistry(in TIME_ZONE_INFORMA
+        /// If not found, uses the preferred default OS UI language, to align with prior behavior.
+        /// </summary>
+        private static string TryGetCachedFileMuiPath(string filePath, CultureInfo cultureInfo)
+        {
+            string? result;
+            int hash = HashCode.Combine(filePath, cultureInfo);
+
+            lock (s_FileMuiPathCache)
+            {
+                if (s_FileMuiPathCache.TryGetValue(hash, out result))
+                {
+                    return result;
+                }
+            }
+
+            result = TryGetFileMuiPath(filePath, cultureInfo);
+
+            lock (s_FileMuiPathCache)
+            {
+                s_FileMuiPathCache.TryAdd(hash, result);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Helper function for getting the MUI path for a given resource and culture.
+        /// Searches the installed OS languages for either an exact matching culture, or one that has the same parent.
+        /// If not found, uses the preferred default OS UI language, to align with prior behavior.
+        /// </summary>
+        private static unsafe string TryGetFileMuiPath(string filePath, CultureInfo cultureInfo)
+        {
+            try
+            {
+                char* fileMuiPath = stackalloc char[Interop.Kernel32.MAX_PATH];
+                char* language = stackalloc char[Interop.Kernel32.LOCALE_NAME_MAX_LENGTH];
+                uint fileMuiPathLength = Interop.Kernel32.MAX_PATH;
+                uint languageLength = Interop.Kernel32.LOCALE_NAME_MAX_LENGTH;
+                ulong enumerator = 0;
+
+                while (true)
+                {
+                    // Search all installed languages.  The enumerator is re-used between loop iterations.
+                    bool succeeded = Interop.Kernel32.GetFileMUIPath(
+                        Interop.Kernel32.MUI_USE_INSTALLED_LANGUAGES,
+                        filePath, language, ref languageLength,
+                        fileMuiPath, ref fileMuiPathLength, ref enumerator);
+
+                    if (!succeeded)
+                    {
+                        // Recurse to search using the parent of the desired culture.
+                        if (cultureInfo.Parent.Name != string.Empty)
+                        {
+                            return TryGetFileMuiPath(filePath, cultureInfo.Parent);
+                        }
+
+                        // Final fallback, using the preferred installed UI language.
+                        enumerator = 0;
+                        succeeded = Interop.Kernel32.GetFileMUIPath(
+                            Interop.Kernel32.MUI_USER_PREFERRED_UI_LANGUAGES,
+                            filePath, language, ref languageLength,
+                            fileMuiPath, ref fileMuiPathLength, ref enumerator);
+
+                        if (succeeded)
+                        {
+                            return new string(fileMuiPath);
```

Comment:
>fileMuiPath [](http://example.com/codeflow?start=46&length=11)

does GetFileMUIPath guarantee ending the string with the null termination? this apply for all buffers used in GetFileMUIPath

Reply:
I'm not sure.  The docs don't make that clear.  I will test and find out.

If not, what's the fastest way to clear the buffers?

Reply:
Nevermind - yes, they do return null-terminated strings, and `new string(char*)` picks it up correctly.

The interesting part with this API is that neither of the length parameters are useful - they stay the same as the buffer size.  (Also, I could never get any good results passing *in* a language, only by enumerating through all languages.)

Reply:
Isn't the length returned in the `pcc` parameters? from the code I am seeing these are in/out paamaters.

```
    // pcchFileMUIPath -
    // [IN/OUT]
    // pcchLanguageBuffer
    // [IN/OUT]    
```

Reply:
Nope.  They always stay at the size of the buffers.  [The docs](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getfilemuipath) talk about how you could call the API twice where the first call gets the size of the required buffer and then the second call gets the actual value using that buffer, but then goes on to say *"The function cannot determine the exact size of the path for all successive calls"*, so the length isn't useful when searching.

Reply:
As a test, I filled both buffers with `'z'` characters before making the call.  After the call, the language buffer was like `"en-US\0zzzzzzzz"` and the mui path was like `"C:\\Windows\\System32\\en-US\\tzres.dll.mui\0zzzzzzzzzz"`  - so the `\0` is being populated correctly on both buffers.

Reply:
ok good. at least set the last character in the buffer to `\0` right before calling `new String()`. this will protects against anything can goes wrong.

Reply:
Done.  I made the buffers one character longer also to accommodate there always being a null char even at max actual length.  I seriously doubt it would ever get there with this particular API's typical values, but you never know.